File: Prácticas/P4/JAVA/Civitas/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.Jugador;
import civitas.OperacionesInmobiliarias;
import civitas.TituloPropiedad;
import civitas.Respuestas;

public class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  public VistaTextual(){
    in = new Scanner (System.in);
  }
 
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar(){
      int opcion = menu("¿Quieres comprar?",
      new ArrayList<> (Arrays.asList("NO","SI")));
      return (Respuestas.values()[opcion]);
  }

  void gestionar(){
      int opcion = menu("¿Que gestion inmobiliaria quieres realizar?",
           new ArrayList<> (Arrays.asList("VENDER","HIPOTECAR","CANCELAR_HIPOTECA","CONSTRUIR_CASA","CONSTRUIR_HOTEL","TERMINAR")));
      iGestion = opcion;
      
      if(iGestion != OperacionesInmobiliarias.TERMINAR.ordinal()){
        ArrayList<String> propiedadesASString = juegoModel.getJugadorActual().cadenaPropiedades();
        int propiedad = menu("¿Sobre que propiedad quieres realizar la operacion?",
                            propiedadesASString);
        iPropiedad = propiedad;
      }
      
  }
  
  public int getGestion(){
      return iGestion;
  }
  
  public int getPropiedad(){
      return iPropiedad;
  }
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion){
      System.out.println("Siguiente operacion " + operacion);
  }


  void mostrarEventos(){
      while(juegoModel.diario.eventosPendientes() == true){
          System.out.println(juegoModel.diario.leerEvento());
      }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();

    }
  
  void actualizarVista(){
      System.out.println("Jugador Actual " + juegoModel.getJugadorActual() + "Casilla actual " + juegoModel.getCasillaActual());
  } 
}

File: Prácticas/P4/RUBY/Civitas/lib/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego'
require 'io/console'
require_relative"operacion_inmobiliaria.rb"
require_relative"respuestas.rb"
require_relative"salidas_carcel.rb"
require_relative"operaciones_juego.rb"
require_relative"civitas_juego.rb"

module Civitas

  class Vista_textual
    
    def initialize
      @juego_model
      @i_gestion = -1
      @i_propiedad = -1
    end

    def mostrar_estado(estado)
      puts estado
    end
    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l.to_s
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opcion: ",
                          tab+"Valor erroneo")
      return opcion
    end
  
    def salir_carcel()
      lista_opciones = [Salidas_carcel::PAGANDO,Salidas_carcel::TIRANDO]
      opcion = menu("Elige la forma para intentar salir de la carcel",lista_opciones)
      return lista_opciones[opcion]
    end

    
    def comprar
      lista_respuestas = [Respuestas::NO,Respuestas::SI]
      
      opcion = menu("¿Quieres comprar?",lista_respuestas)
      
      return lista_respuestas[opcion]
    end

    def gestionar
      
      lista_operaciones_inmobiliarias = [Operaciones_inmobiliarias::VENDER,Operaciones_inmobiliarias::HIPOTECAR,Operaciones_inmobiliarias::CANCELAR_HIPOTECA,Operaciones_inmobiliarias::CONSTRUIR_CASA,Operaciones_inmobiliarias::CONSTRUIR_HOTEL,Operaciones_inmobiliarias::TERMINAR]
      opcion = menu("¿Que gestion inmobiliaria quieres realizar?", lista_operaciones_inmobiliarias)
      @i_gestion = opcion
      
      puts "IGESTION 0 " + @i_gestion.to_s
      
      if(@i_gestion != 5)
        propiedadesASString = @juego_model.get_jugador_actual.cadena_propiedades
        propiedad = menu("¿Sobre que propiedad quieres realizar la operacion?",propiedadesASString)
        @i_propiedad = propiedad
      end
      
    end

    def get_gestion
      return @i_gestion
    end

    def get_propiedad
      return @i_propiedad
    end

    def mostrar_siguiente_operacion(operacion)
      puts "Siguiente operacion #{operacion}"
    end

    def mostrar_eventos
      while(@juego_model.get_diario.eventos_pendientes == true)
        puts @juego_model.get_diario.leer_evento
      end
    end

    def set_civitas_juego(civitas)
         @juego_model=civitas
         self.actualizar_vista
    end

    def actualizar_vista
      puts "Jugador actual \n #{@juego_model.get_jugador_actual} \n Casilla actual #{@juego_model.get_casilla_actual}"
    end

    
  end

end
